### PR TITLE
[#33] Audit contract

### PR DESCRIPTION
**Pull request by pafff**

### CODE REVIEW

File: clove/network/bitcoin.py [hunk 1]
```diff
@@ -219,6 +221,55 @@ def show_details(self):
...
+            self.recipient_address = P2PKHBitcoinAddress.from_bytes(script_ops[6])
+            self.refund_address = P2PKHBitcoinAddress.from_bytes(script_ops[13])
+            self.locktime = datetime.fromtimestamp(int(script_ops[8]))
+            self.secret_hash = b2x(script_ops[2])
+            self.value = contract_tx_out.nValue / COIN
+        else:
+            raise ValueError('Given transaction is not a valid contract.')
+
+    @staticmethod
+    def is_valid_contract_script(script_ops):
```
Comment (szerepak) [inline clove/network/bitcoin.py:248]: :eyes: these asserts within try-except look bizarre
I would go with assigning that to variable and returning that.

File: clove/network/bitcoin.py [hunk 2]
```diff
@@ -219,6 +221,55 @@ def show_details(self):
...
+            raise ValueError('Given transaction has no outputs.')
+
+        contract_tx_out = self.tx.vout[0]
+        script_ops = list(contract_tx_out.scriptPubKey)
+        if self.is_valid_contract_script(script_ops):
+            self.recipient_address = P2PKHBitcoinAddress.from_bytes(script_ops[6])
+            self.refund_address = P2PKHBitcoinAddress.from_bytes(script_ops[13])
+            self.locktime = datetime.fromtimestamp(int(script_ops[8]))
+            self.secret_hash = b2x(script_ops[2])
+            self.value = contract_tx_out.nValue / COIN
```
Comment (szerepak) [inline clove/network/bitcoin.py]: hmm, maybe it's worth to make a util method for this specific division with a good naming?

**Review by szerepak — APPROVED** [review]

File: clove/utils/bitcoin.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from bitcoin.core import COIN
+
+
+def satoshi_to_btc(value):
+    return value / COIN
+
+
+def btc_to_satoshi(value):
```
Comment (trojkat) [inline clove/utils/bitcoin.py:8]: 👍

File: clove/utils/bitcoin.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from bitcoin.core import COIN
+
+
+def satoshi_to_btc(value):
```
Comment (trojkat) [inline clove/utils/bitcoin.py:4]: 👍

File: tests/test_bitcoin_transfer.py [hunk 5]
```diff
@@ -95,3 +96,40 @@ def test_transaction_fee(fee_per_kb_mock, unsigned_transaction):
...
+
+def test_audit_contract(signed_transaction):
+    btc_network = BitcoinTestNet()
+    transaction_details = signed_transaction.show_details()
+
+    contract = btc_network.audit_contract(transaction_details['contract_transaction'])
+    contract_details = contract.show_details()
+
+    assert contract.secret_hash == transaction_details['secret_hash']
+    assert contract.value == signed_transaction.value
```
Comment (trojkat) [inline tests/test_bitcoin_transfer.py]: I would match field to field and details to details.

File: tests/test_bitcoin_transfer.py [hunk 6]
```diff
@@ -95,3 +96,40 @@ def test_transaction_fee(fee_per_kb_mock, unsigned_transaction):
...
+    assert contract.locktime == signed_transaction.locktime.replace(microsecond=0)
+    assert contract_details['recipient_address'] == signed_transaction.recipient_address
+    assert contract_details['refund_address'] == signed_transaction.sender_address
+    assert contract_details['value_text'] == transaction_details['value_text']
+    assert contract_details['transaction_hash'] == transaction_details['contract_transaction_hash']
```
Comment (trojkat) [inline tests/test_bitcoin_transfer.py]: Maybe lets rename `contract_transaction_hash` also to `transaction_hash`

**Review by trojkat — APPROVED** [review]

**Pull request merged**